File: matlab/run_aerostruct.m
% Similar to run_aerostruct.py example script

try
    % on Unix systems this setting is required otherwise Matlab crashes
    py.sys.setdlopenflags(int32(10));  % Set RTLD_NOW and RTLD_DEEPBIND
catch
end

% load Python from virtual environment with OpenMDAO 1.7.3 installed
fprintf('Load Python... \n')
[~,~,isloaded] = pyversion;
if ~isloaded
   pyversion 'C:\Users\Sam\repos\OpenAeroStruct\venv\Scripts\python.exe'
end

% import OpenAeroStruct python module
% OAS_PATH = '/general/home/samfriedman';
OAS_PATH = 'C:\Users\samfriedman\repos\OpenAeroStruct';
P = py.sys.path;
if count(P,OAS_PATH) == 0
    insert(P,int64(0),OAS_PATH);
end
fprintf('Import OpenAeroStruct module... \n');
py.importlib.import_module('OpenAeroStruct');

prob_dict = struct;
prob_dict.type = 'aerostruct';
prob_dict.with_viscous = true;
prob_dict.cg = mat2np([30., 0., 5.]);
% prob_dict.cg = mat2np([30., 0., 5.]);
prob_dict.optimize = false;
prob_dict.record_db = false;  % using sqlitedict locks a process
prob_dict.print_level = 0;

% Instantiate problem and add default surface
fprintf('Create OASProblem object with prob_dict... \n');
OAS_prob = py.OpenAeroStruct.run_classes.OASProblem(prob_dict);

% Create a dictionary to store options about the surface
surf_dict = struct;
surf_dict.name = 'wing';
surf_dict.num_y = int64(7);
surf_dict.num_x = int64(2);
surf_dict.wing_type = 'CRM';
surf_dict.CD0 = 0.015;
surf_dict.symmetry = true;
surf_dict.num_twist_cp = int64(2);
surf_dict.num_thickness_cp = int64(2);

% Add the specified wing surface to the problem
fprintf('Add wing surface to problem... \n');
OAS_prob.add_surface(surf_dict);

%% Add design variables, constraint, and objective on the problem
OAS_prob.add_desvar('alpha', pyargs('lower',-10., 'upper',10.));
OAS_prob.add_constraint('L_equals_W', pyargs('equals', 0.));
OAS_prob.add_objective('fuelburn', pyargs('scaler', 1e-5));

% Multiple lifting surfaces
surf_dict = struct;
surf_dict.name = 'tail';
surf_dict.num_y = int64(7);
surf_dict.num_x = int64(2);
surf_dict.span = 20.;
surf_dict.root_chord = 5.;
surf_dict.wing_type = 'rect';
surf_dict.offset = [50., 0., 5.];
surf_dict.twist_cp = -9.5;
% surf_dict.root_chord = 5.;
% surf_dict.wing_type = 'rect';
% surf_dict.offset = mat2np([50., 0., 5.]);
% surf_dict.twist_cp = mat2np(-9.5); %#ok<NBRAK>
%
fprintf('Add tail surface to problem... \n');
OAS_prob.add_surface(surf_dict)

% Add design variables and constraints for both the wing and tail
fprintf('Add design variables and constraints... \n');
OAS_prob.add_desvar('wing.twist_cp', pyargs('lower',-15.,'upper',15.));
OAS_prob.add_desvar('wing.thickness_cp', pyargs('lower',0.01, 'upper',0.5, 'scaler',1e2));
OAS_prob.add_constraint('wing_perf.failure', pyargs('upper',0.));
OAS_prob.add_constraint('wing_perf.thickness_intersects', pyargs('upper',0.));
OAS_prob.add_desvar('tail.twist_cp', pyargs('lower',-15., 'upper',15.));
OAS_prob.add_desvar('tail.thickness_cp', pyargs('lower',0.01,'upper',0.5,'scaler',1e2));
OAS_prob.add_constraint('tail_perf.failure', pyargs('upper',0.));
OAS_prob.add_constraint('tail_perf.thickness_intersects', pyargs('upper',0.));

% Setup problem
fprintf('Set up the problem... \n');
OAS_prob.setup()

% Actually run the problem
fprintf('Run the problem... \n');
tic;
output = struct(OAS_prob.run(pyargs('matlab',true)));
t = toc;
fuelburn = output.fuelburn;

fprintf('\nFuelburn: %.4f \n', fuelburn);
fprintf('Time elapsed: %.4f secs\n', t);

File: run_aerostruct.py
""" Example runscript to perform aerostructural optimization.

Call as `python run_aerostruct.py 0` to run a single analysis, or
call as `python run_aerostruct.py 1` to perform optimization.

Call as `python run_aerostruct.py 0m` to run a single analysis with
multiple surfaces, or
call as `python run_aerostruct.py 1m` to perform optimization with
multiple surfaces.

"""

from __future__ import division, print_function
import sys
from time import time
import numpy as np

# Append the parent directory to the system path so we can call those Python
# files. If you have OpenAeroStruct in your PYTHONPATH, this is not necessary.
from os import sys, path
sys.path.append(path.dirname(path.dirname(path.abspath(__file__))))

from OpenAeroStruct import OASProblem

# Suppress warnings
import warnings
warnings.filterwarnings("ignore")

if __name__ == "__main__":

    # Make sure that the user-supplied input is one of the valid options
    input_options = ['0', '0m', '1', '1m']
    print_str = ''.join(str(e) + ', ' for e in input_options)

    # Parse the user-supplied command-line input and store it as input_arg
    try:
        input_arg = sys.argv[1]
        if input_arg not in input_options:
            raise(IndexError)
    except IndexError:
        print('\n +---------------------------------------------------------------+')
        print(' | ERROR: Please supply a correct input argument to this script. |')
        print(' | Possible options are ' + print_str[:-2] + '                             |')
        print(' | See the docstring at the top of this file for more info.      |')
        print(' +---------------------------------------------------------------+\n')
        raise


    # Set problem type
    prob_dict = {'type' : 'aerostruct',
                 'with_viscous' : True,
                 'alpha': 0.,
                 'print_level': 0
                 # 'cg' : np.array([30., 0., 5.])
                 }

    if sys.argv[1].startswith('0'):  # run analysis once
        prob_dict.update({'optimize' : False})
    else:  # perform optimization
        prob_dict.update({'optimize' : True})

    # Instantiate problem and add default surface
    OAS_prob = OASProblem(prob_dict)

    # Create a dictionary to store options about the surface
    surf_dict = {'num_y' : 7,
                 'num_x' : 2,
                 'wing_type' : 'CRM',
                 'CD0' : 0.015,
                 'symmetry' : True,
                 'num_twist_cp' : 2,
                 'num_thickness_cp' : 2,
                 'num_chord_cp': 1,
                 'exact_failure_constraint': True,
                 'span_cos_spacing': 0.5}

    # Add the specified wing surface to the problem
    OAS_prob.add_surface(surf_dict)

    # Add design variables, constraint, and objective on the problem
    # OAS_prob.add_desvar('alpha', lower=-10., upper=10.)
    OAS_prob.add_constraint('L_equals_W', equals=0.)
    OAS_prob.add_objective('fuelburn', scaler=1e-5)

    # Single lifting surface
    if not sys.argv[1].endswith('m'):

        # Setup problem and add design variables, constraint, and objective
        OAS_prob.add_desvar('wing.twist_cp', lower=-15., upper=15.)
        OAS_prob.add_desvar('wing.thickness_cp', lower=0.01, upper=0.5, scaler=1e2)
        OAS_prob.add_constraint('wing_perf.failure', upper=0.)
        OAS_prob.add_constraint('wing_perf.thickness_intersects', upper=0.)
        OAS_prob.setup()

    # Multiple lifting surfaces
    else:

        # Add additional lifting surface
        surf_dict = {'name' : 'tail',
                     'num_y' : 7,
                     'num_x' : 2,
                     'span' : 20.,
                     'root_chord' : 5.,
                     'wing_type' : 'rect',
                     'offset' : np.array([50., 0., 5.]),
                     'twist_cp' : np.array([-9.5]),
                     'exact_failure_constraint': True}
        OAS_prob.add_surface(surf_dict)

        # Add design variables and constraints for both the wing and tail
        OAS_prob.add_desvar('wing.twist_cp', lower=-15., upper=15.)
        OAS_prob.add_desvar('wing.thickness_cp', lower=0.01, upper=0.5, scaler=1e2)
        OAS_prob.add_desvar('wing.chord_cp', lower=0.9, upper=1.1)
        OAS_prob.add_desvar('wing.taper', lower=0.2, upper=1.1)
        OAS_prob.add_constraint('wing_perf.failure', upper=0.)
        OAS_prob.add_constraint('wing_perf.thickness_intersects', upper=0.)
        # OAS_prob.add_desvar('tail.twist_cp', lower=-15., upper=15.)
        # OAS_prob.add_desvar('tail.thickness_cp', lower=0.01, upper=0.5, scaler=1e2)
        # OAS_prob.add_constraint('tail_perf.failure', upper=0.)
        # OAS_prob.add_constraint('tail_perf.thickness_intersects', upper=0.)

        # Setup problem
        OAS_prob.setup()

    st = time()
    # Actually run the problem
    if not prob_dict['optimize']:
        thickness_cp = np.array([
            [0.03709672, 0.04913425],
            [0.03, 0.05],
            [0.02, 0.05],
            [0.03777685, 0.07183272]
        ])
        twist_cp = np.array([
            [5.65946593, -3.53034877],
            [5, -3],
            [0, 0],
            [12.80374032, 14.73784563]
        ])
        taper = np.array([
            [0.2],
            [1.0],
            [0.75],
            [0.2]
        ])
        chord_cp = np.array([
            [0.9],
            [1.0],
            [1.1],
            [0.9]
        ])
        out = []
        for i in range(thickness_cp.shape[0]):
            input_dict = {
                'wing.thickness_cp': thickness_cp[i],
                'wing.twist_cp': twist_cp[i],
                'wing.taper': taper[i],
                'wing.chord_cp': chord_cp[i]
            }
            out.append(OAS_prob.run(**input_dict))
            print(' ---  NEW RUN  ---')
            for key, val in out[i].iteritems():
                print(key, val)
    else:
        out = OAS_prob.run()
        for key, val in out.iteritems():
            print(key, val)

    # for key, val in OAS_prob.prob.root._unknowns_dict.iteritems():
    #     print(val['top_promoted_name'],' --> ',key)
    #
    #

    print("\nFuelburn:", OAS_prob.prob['fuelburn'])
    print("Time elapsed: {} secs".format(time() - st))
